File: minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "": return ""

        target = {}
        window = {}

        for ch in t:
            target[ch] = 1 + target.get(ch, 0)

        have, need = 0, len(target)
        ans, length = [-1, -1], float("inf")
        left = 0

        for right in range(len(s)):
            ch = s[right]
            window[ch] = 1 + window.get(ch, 0)

            if ch in target and window[ch] == target[ch]:
                have += 1

            while have == need:
                if (right-left+1) < length:
                    ans = [left, right]
                    length = right - left + 1

                window[s[left]] -= 1
                if s[left] in target and window[s[left]] < target[s[left]]:
                    have -= 1
                left += 1

        left, right = ans
        return s[left:right+1] if length != float("inf") else ""

File: minimum-window-substring.cpp
#include <string>
#include <climits>
#include <unordered_map>

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() == 0) return t;

		unordered_map<char, int> target, window;

		for (char ch: t) {
			target[ch] = (target.find(ch) != target.end())? target[ch]+1: 1;
		}
		
		int have = 0;
		int need = target.size();
		int ileft = -1;
		int iright = -1;
		int length = INT_MAX;
		int left = 0;

		for (int right=0; right<s.size(); ++right) {
			char ch = s[right];
			window[ch] = (window.find(ch) != window.end())? window[ch]+1: 1;

			if ((target.find(ch) != target.end()) && (window[ch] == target[ch])) {
				have += 1;
			}

			while (have == need) {
				if ((right-left+1) < length) {
					ileft = left;
					iright = right;
					length = right - left + 1;
				}
				
				window[s[left]] -= 1;
				char ch = s[left];
				
				if ((target.find(ch) != target.end()) && (window[ch] < target[ch])) {
					have -= 1;
				}
				
				left += 1;
			}
		}

		return (length != INT_MAX)? s.substr(ileft, length): "";
    }
};
